precompiles/xcm/src/lib.rs

```diff
@@ -0,0 +1,160 @@
+{
+    fn execute(
+        input: &[u8], //Reminder this is big-endian
+        target_gas: Option<u64>,
+        context: &Context,
+        is_static: bool,
+    ) -> EvmResult<PrecompileOutput> {
+        log::trace!(target: "xcm-precompile", "In XCM precompile");
+
+        let mut gasometer = Gasometer::new(target_gas);
+        let gasometer = &mut gasometer;
+
+        let (mut input, selector) = EvmDataReader::new_with_selector(gasometer, input)?;
+        let input = &mut input;
+
+        gasometer.check_function_modifier(context, is_static, FunctionModifier::NonPayable)?;
+
+        match selector {
+            // Dispatchables
+            Action::AssetsWithdraw => Self::assets_withdraw(input, gasometer, context),
+        }
+    }
+}
+
+impl<R, C> XcmPrecompile<R, C>
+where
+    R: pallet_evm::Config
+        + pallet_xcm::Config
+        + pallet_assets::Config
+        + AddressToAssetId<<R as pallet_assets::Config>::AssetId>,
+    <<R as frame_system::Config>::Call as Dispatchable>::Origin: From<Option<R::AccountId>>,
+    <R as frame_system::Config>::Call:
+        From<pallet_xcm::Call<R>> + Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+    C: Convert<MultiLocation, <R as pallet_assets::Config>::AssetId>,
+{
+    fn assets_withdraw(
+        input: &mut EvmDataReader,
+        gasometer: &mut Gasometer,
+        context: &Context,
+    ) -> EvmResult<PrecompileOutput> {
+        // Bound check
+        input.expect_arguments(gasometer, 6)?;
+
+        // Read arguments and check it
+        let assets: Vec<MultiLocation> = input
+            .read::<Vec<Address>>(gasometer)?
+            .iter()
+            .cloned()
+            .filter_map(|address| {
+                R::address_to_asset_id(address.into())
+                    .map(|x| C::reverse_ref(x).ok())
+                    .flatten()
+            })
+            .collect();
+        let amounts: Vec<u128> = input
+            .read::<Vec<U256>>(gasometer)?
+            .iter()
+            .map(|x| x.low_u128())
+            .collect();
+
+        // Check that assets list is valid:
+        // * all assets resolved to multi-location
+        // * all assets has corresponded amount
+        if assets.len() != amounts.len() || assets.len() == 0 {
+            return Err(gasometer.revert("assets resolution failure"));
+        }
+
+        let recipient: [u8; 32] = input.read::<H256>(gasometer)?.into();
+        let is_relay = input.read::<bool>(gasometer)?;
+        let parachain_id: u32 = input.read::<U256>(gasometer)?.low_u32();
+        let fee_asset_item: u32 = input.read::<U256>(gasometer)?.low_u32();
+
+        if fee_asset_item as usize > assets.len() {
+            return Err(gasometer.revert("bad fee index"));
+        }
+
+        // Prepare pallet-xcm call arguments
+        let dest = if is_relay {
+            MultiLocation::parent()
+        } else {
+            X1(Junction::Parachain(parachain_id)).into()
+        };
```

Comment:
Tbh, I have some trouble wrapping my head around how useful is this part.

Based on the provided asset Ids, we know we exactly where we need to deposit them back.
If we don't, it means the asset mapping is missing and something's wrong.

Would it be valid to just take the first destination location derived from asset Id and ensure that it's the same parachain for all assets in the list?

Reply:
You're right, I forgot about how I did it in demo.

Westmint accepted `{parents: 1, interior: Here}` as _native_ currency.